src/ripple/rpc/RPCVersion.h

```diff
@@ -0,0 +1,46 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2012, 2013 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLED_RIPPLE_RPC_VERSIONS_H
+#define RIPPLED_RIPPLE_RPC_VERSIONS_H
+
+#include <beast/module/core/diagnostic/SemanticVersion.h>
+#include <ripple/protocol/JsonFields.h>
+#include <ripple/rpc/impl/JsonObject.h>
+
+namespace ripple {
+namespace RPC {
+
+extern beast::SemanticVersion const firstVersion;
+extern beast::SemanticVersion const bestVersion;
+extern beast::SemanticVersion const lastVersion;
+
+template <class Object>
+void setVersion(Object& parent)
+{
+    auto&& object = addObject (parent, jss::version);
```

Comment:
So I've done this quite a bit already and the `&&` is essential for this generic code.

We now have two types of Json things - we have a mutable, hashed structure `Json::Value` where you can set and reset parts, do surgery and that sort of thing.  This is convenient but unfortunately it's big and slow and its parts show up all over our profiling.

We also have a write-only JSON class called `RPC::JsonObject` (soon to be `Json::Object`) which directly outputs JSON as it goes, avoiding all those little news and deletes.  Json::Object can also be set up to yield periodically so we can keep rescheduling large jobs onto the job queue instead of having the job queue essentially hang for up to a minute while it prepares a huge response.

All this code works on both sorts of objects.  There are some cleanups coming to make the interoperation even more effortless.

`addObject` returns a reference in the first case, but an rvalue in the second case, so the `&&` is unavoidable.  Not a big deal though, it works well!